dg.go

```diff
@@ -244,14 +244,16 @@ func (d *directedGraph[NodeType]) HasReadyNodes() bool {
 }
 
 func (d *directedGraph[NodeType]) PopReadyNodes() map[string]ResolutionStatus {
+	result := make(map[string]ResolutionStatus)
 	d.lock.Lock()
+	defer d.lock.Unlock()
 	// Transfer the map to a local variable to minimize time locked, and reset the graph's value.
 	readyMap := d.readyForProcessing
 	d.readyForProcessing = make(map[string]*node[NodeType])
-	d.lock.Unlock()
 
-	result := make(map[string]ResolutionStatus)
 	for _, node := range readyMap {
+		// Technically, while unlikely, it's possible for the node's status to race with
+		// this access of node.status.
```

Comment:
I don't think this comment is helpful (for about six different reasons).  What is it that you're really trying to say here?

Once this function returns and the lock is unlocked, it's possible that any of the returned nodes' statuses could change, but the caller will never know (and shouldn't care) unless it calls in again to check.

Moreover, the nodes in this list are _supposed_ to be in terminal states such that their statuses will not be changed again.  (Is that correct??)

So, where is the "race"?  (Go's "race detection" aside, a _race condition_ only results if, owing to a lack of synchronization, the result of an operation is dependent on an unreliable order of operations -- typically, this requires selecting a branch based on a shared value -- so, where is the branch?)

Reply:
I updated the comment to specify the scenario.

Reply:
The new comment is worlds better than the previous, but....

- Can a node's status ever be changed without holding the lock?
- Nodes on the ready list are always in either the `Waiting` or `Unresolvable` state, right?
- Nodes _on the ready list_ cannot become `Resolved`, and they cannot transition from `Waiting` to `Unresolvable` (because their inputs are already satisfied), right?

So, if a node on the ready list is the target of a status change, it can only be `Waiting -> Waiting` or `Unresolvable -> Unresolvable` right??  (Before a node goes from `Waiting -> Resolved` it would have had to have been removed from the ready list first...no?)

So, what is triggering Go's race detector is not actually a problem:  it's the fact that we are (were) reading a node without the lock when at other times we write it with the lock.  This is the base requirement for a race condition (which is why Go reports it), but it doesn't actually result in a race condition in this particular case.

Yes, we can prevent Go's race detector from triggering here by reading the node under the lock, but that isn't to say that "we must do this under the lock" unless you add "to avoid triggering Go's race detector"!

So, I recommend that you simply delete this comment.  I don't think it improves the understanding of this code.  As the code is now structured, we need to hold the lock to guard the accesses to `d.readyForProcessing` and we get coverage for our accesses to `node` "for free".  Trying to say anything more just muddies the waters, because it doesn't matter whether whether there are things which concurrently write the `status` value, because they are not going to change it in any way that matters.

Reply:
> Can a node's status ever be changed without holding the lock?

No. But this code accessed it without the lock, causing the problem.
If a user decides to mark a node as unresolvable in another thread, that will lock the mutex.

> Nodes on the ready list are always in either the Waiting or Unresolvable state, right?

They could be in any state. Including Resolved if the node is resolved before it's marked ready.

> Nodes on the ready list cannot become Resolved, and they cannot transition from Waiting to Unresolvable (because their inputs are already satisfied), right?

As stated in the above response, they can show up as resolved. But you cannot transition from Unresolvable to anything, or Resolved to anything. It's just waiting to one of those.
The users can choose to mark one of those nodes as unresolvable depending on how they're using them. It's up to them to determine if that design is wrong.

---

So it is very much by design to allow the users of the dgraph to mark nodes as unresolvable without involving PopReadyNodes, and to do so on another thread. So this needs to be locked for that, and not for just satisfying the Go Race Detector.